proxymode/resolver.go

```diff
@@ -16,25 +17,57 @@ type proxyResolver struct {
 	destinationURL url.URL
 }
 
+var acceptedHeaders = map[string]string{
+	"Content-Length": "",
+	"Content-Type":   "",
+	"Last-Modified":  "",
+}
+
+func (pr proxyResolver) pipeRequestProxy(w http.ResponseWriter, remotePath string) error {
+	remoteURL := pr.destinationURL.ResolveReference(&url.URL{Path: remotePath})
+
+	resp, err := http.Get(remoteURL.String())
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	for header, values := range resp.Header {
+		if len(w.Header().Values(header)) > 0 {
+			continue
+		}
+		if _, ok := acceptedHeaders[header]; !ok {
+			continue
+		}
+		for _, value := range values {
+			w.Header().Add(header, value)
+		}
+	}
+	w.WriteHeader(resp.StatusCode)
+
+	_, err = io.Copy(w, resp.Body)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
 func (pr proxyResolver) RedirectArtifactsHandler(w http.ResponseWriter, r *http.Request, p *packages.Package) {
 	remotePath := fmt.Sprintf("/epr/package/%s-%s.zip", p.Name, p.Version)
-	anURL := pr.destinationURL.
-		ResolveReference(&url.URL{Path: remotePath})
-	http.Redirect(w, r, anURL.String(), http.StatusMovedPermanently)
+
+	pr.pipeRequestProxy(w, remotePath)
 }
 
 func (pr proxyResolver) RedirectStaticHandler(w http.ResponseWriter, r *http.Request, p *packages.Package, resourcePath string) {
 	remotePath := fmt.Sprintf("/package/%s/%s/%s", p.Name, p.Version, resourcePath)
-	staticURL := pr.destinationURL.
-		ResolveReference(&url.URL{Path: remotePath})
-	http.Redirect(w, r, staticURL.String(), http.StatusMovedPermanently)
+
+	pr.pipeRequestProxy(w, remotePath)
 }
 
 func (pr proxyResolver) RedirectSignaturesHandler(w http.ResponseWriter, r *http.Request, p *packages.Package) {
```

Comment:
I was thinking in options to rename them, I could rename them to be "ForwardArtifactsHandler, ForwardStaticHandler,..." for instance.

We should keep a generic name because in proxy mode it would be interesting to keep the `http.Redirect` instead, wouldn't it ? In following changes, I was about to move this implementation just to `storage/resolver.go`

WDYT ?

Reply:
Ok, sounds good to me.